core/offchain/src/api.rs

```diff
@@ -52,38 +69,122 @@ fn unavailable_yet<R: Default>(name: &str) -> R {
 
 const LOCAL_DB: &str = "LOCAL (fork-aware) DB";
 const STORAGE_PREFIX: &[u8] = b"storage";
+const KEYS_PREFIX: &[u8] = b"keys";
+
+const NEXT_ID: &[u8] = b"crypto_key_id";
+
+impl<Storage, KeyProvider> Api<Storage, KeyProvider> where
+	Storage: OffchainStorage,
+	KeyProvider: AuthorityKeyProvider,
+{
+	fn keypair<P: Pair>(&self, phrase: &str) -> Result<P, ()> {
+		P::from_phrase(phrase, Some(self.keys_password.as_ref()))
+			.map_err(|e| {
+				warn!("Error recovering Offchain Worker key. Password invalid? {:?}", e);
+				()
+			})
+			.map(|x| x.0)
+	}
+
+	fn read_key(&self, id: Option<CryptoKeyId>, kind: CryptoKind) -> Result<Key, ()> {
+		if let Some(id) = id {
+			let key = self.db.get(KEYS_PREFIX, &id.0.encode())
+				.and_then(|key| CryptoKey::decode(&mut &*key))
+				.ok_or(())?;
+			if key.kind != kind {
+				warn!(
+					"Invalid crypto kind (got: {:?}, expected: {:?}), when requesting key {:?}",
+					key.kind,
+					kind,
+					id
+				);
+				return Err(())
+			}
+
+			Ok(match key.kind {
+				CryptoKind::Sr25519 => Key::Sr25519(self.keypair(&key.phrase)?),
+				CryptoKind::Ed25519 => Key::Ed25519(self.keypair(&key.phrase)?),
+			})
+		} else {
+			let key = match kind {
+				CryptoKind::Sr25519 => self.key_provider.authority_key().map(Key::Sr25519),
+				CryptoKind::Ed25519 => self.key_provider.authority_key().map(Key::Ed25519),
+			};
+
+			key.ok_or_else(|| {
+				warn!("AuthorityKey is not configured, yet offchain worker tried to access it.");
+				()
+			})
+		}
+	}
+}
 
-impl<S: OffchainStorage> OffchainExt for Api<S> {
+impl<Storage, KeyProvider> OffchainExt for Api<Storage, KeyProvider> where
+	Storage: OffchainStorage,
+	KeyProvider: AuthorityKeyProvider,
+{
 	fn submit_transaction(&mut self, ext: Vec<u8>) -> Result<(), ()> {
 		self.sender
 			.unbounded_send(ExtMessage::SubmitExtrinsic(ext))
 			.map(|_| ())
 			.map_err(|_| ())
 	}
 
-	fn new_crypto_key(&mut self, _crypto: CryptoKind) -> Result<CryptoKeyId, ()> {
-		unavailable_yet::<()>("new_crypto_key");
-		Err(())
+	fn new_crypto_key(&mut self, kind: CryptoKind) -> Result<CryptoKeyId, ()> {
+		let phrase = match kind {
+			CryptoKind::Ed25519 => {
+				ed25519::Pair::generate_with_phrase(Some(self.keys_password.as_ref())).1
+			},
+			CryptoKind::Sr25519 => {
+				sr25519::Pair::generate_with_phrase(Some(self.keys_password.as_ref())).1
+			},
+		};
+
+		let (id, id_encoded) = loop {
+			let encoded = self.db.get(KEYS_PREFIX, NEXT_ID);
+			let encoded_slice = encoded.as_ref().map(|x| x.as_slice());
+			let new_id = encoded_slice.and_then(|mut x| u16::decode(&mut x)).unwrap_or_default()
```

Comment:
I think we can end here in an infinite loop. If `db.get` returns nothing or `decode` is failing, `unwrap_or_default()` is returning `0` and no other check will end the loop.

Reply:
Hmm, so `compare_and_set` should fail only if the underlying value was modified in the meantime. So even if the value is invalid (and `decode` failed) or if the value is not set (`unwrap_or_default`) the `compare_and_set` should still succeed and we should break the loop.

I can add an upper bound here to prevent infinite loop, but don't really have any idea how we could avoid it.

Reply:
Ahh yeah.